_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+the browsers & devices we support on [artsy.net](artsy.net), but it's the data we need to optimize our responsive
+rendering. We're using it to determine if your browser is likely going to use only the mobile breakpoint of our
+app, in which case we don't have to also render the desktop version.
+
+We aren't doing any detection of desktop browsers. Maybe this is something we do in the future, but for now we are
+more concerned with mobile users getting less content sent over their 3G connection.
+
+## Why didn't you use \_\_\_?
+
+Those are our primary tools for combining SSR with responsive design! They work well for us. We considered many
+many other options along the way. Here are a couple:
+
+### [`react-media`](https://github.com/ReactTraining/react-media) or [`react-responsive`]()
+
+We investigated both `react-media` and `react-responsive`, but found that they didn't approach the SSR side of the
+problem as deeply as we needed.
+
+We also weren't fans of the imperative API in `react-media`. We started with a similar API when building
+`react-responsive-media`, but learned that it limited our ability to render all breakpoints from the server. The
+API requires only one branch to be rendered, and this caused the aforementioned mismatches when the app would
+hydrate.
+
+With `react-responsive`, we didn't like that it relied on user agent detection as its primary method of handling
+SSR.
+
+(TODO: are those statements correct?)
+
+### CSS
+
+We explored the idea of rendering all breakpoints from the server, and hiding the non-matching branches with CSS.
+The issue with this approach is that you have many components that are mounted & rendered unnecessarily. Aside from
+the performance hit you take for rendering components your user isn't seeing, even worse is the potential for
+duplicate side-effects.
+
+Imagine a component that, when rendered, emits a call to an analytics service. If this component exists in both a
+mobile and desktop branch, you're now double-stuffing your analytics. Hopefully your analytics service is smart
+enough to count both calls, but it's still a bad idea to duplicate components that have side-effects.
+
+## What's left to solve?
+
+Our SSR & responsive design toolbox does a lot of things well. We get great performance from both the server and
+client. Our site looks great on any device.
+
+We do have some SEO concerns, though. Since we're server-rendering multiple breakpoints, it's likely that search
+engine bots are seeing double the content on our pages. We _think_ this is okay.
+[Google WebMasters](https://youtu.be/WsgrSxCmMbM) says it's okay. We haven't noticed any awful side-effects from
+this yet, but SEO is a bit of a dark art, yeah?
+
+We _have_ had some issues where Google is showing limited data for some of our pages. We don't know yet if it is
+related to truncation we're doing at the mobile breakpoint, but we're investigating!
```

Comment:
Maybe elaborate on this or remove completely?

Reply:
I'll think about it!

Reply:
Removed!